firewall/firewall_test.go

```diff
@@ -0,0 +1,246 @@
+		"INSERT INTO Customers (CustomerName, City, Country) VALUES ('Cardinal', 'Stavanger', 'Norway');",
+		//"INSERT INTO SalesStaff1 VALUES (1, 'Stephen', 'Jiang');",
+		"INSERT SalesStaff1 VALUES (2, 'Michael', 'Blythe'), (3, 'Linda', 'Mitchell'),(4, 'Jillian', 'Carson'), (5, 'Garrett', 'Vargas');",
+		"INSERT INTO SalesStaff2 (StaffGUID, FirstName, LastName) VALUES (NEWID(), 'Stephen', 'Jiang');",
+		"INSERT INTO Customers (CustomerName, ContactName, Address, City, PostalCode, Country) VALUES ('Cardinal', 'Tom B. Erichsen', 'Skagen 21', 'Stavanger', '4006', 'Norway');",
+		"INSERT INTO Customers (CustomerName, City, Country) VALUES ('Cardinal', 'Stavanger', 'Norway');",
+		"INSERT INTO films VALUES ('UA502', 'Bananas', 105, '1971-07-13', 'Comedy', '82 minutes');",
+		"INSERT INTO films (code, title, did, date_prod, kind) VALUES ('B6717', 'Tampopo', 110, '1985-02-10', 'Comedy'), ('HG120', 'The Dinner Game', 140, DEFAULT, 'Comedy');",
+		"INSERT INTO films SELECT * FROM tmp_films WHERE date_prod < '2004-05-07';",
+	}
+
+	blackList := [] string {
+		"INSERT INTO SalesStaff1 VALUES (1, 'Stephen', 'Jiang');",
+		"SELECT AVG(Price) FROM Products;",
+	}
+
+
+	blacklistHandler := &handlers.BlacklistHandler{}
+	blacklistHandler.AddQueries(blackList)
+
+	firewall := &Firewall{}
+
+	var err error
+
+	//set our firewall to use blacklist for query evaluating
+	firewall.AddHandler(blacklistHandler)
+
+	for _, query := range sqlSelectQueries{
+		err = firewall.HandleQuery(query)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	for _, query := range sqlInsertQueries{
+		err = firewall.HandleQuery(query)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	testQuery := "INSERT INTO Customers (CustomerName, City, Country) VALUES ('Cardinal', 'Stavanger', 'Norway');";
+
+	blacklistHandler.AddQueries([]string{testQuery})
+
+	err = firewall.HandleQuery(testQuery)
+	//firewall should block this query by throwing error
+	if err == nil {
+		t.Fatal(err)
+	}
+
+	firewall.RemoveHandler(blacklistHandler)
+
+	err = firewall.HandleQuery(testQuery)
+	//firewall should not block this query because we removed blacklist handler, err should be nil
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	//again set our firewall to use blacklist for query evaluating
+	firewall.AddHandler(blacklistHandler)
+
+	//now firewall should block testQuery by throwing error
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	blacklistHandler.RemoveQueries([]string{testQuery})
+
+	err = firewall.HandleQuery(testQuery)
+	//now firewall should not block testQuery
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testBlacklistTables(t, firewall, blacklistHandler)
+}
+
+func testBlacklistTables(t *testing.T, firewall * Firewall, blacklistHandler * handlers.BlacklistHandler){
+
+	testQueries := []string {
+		"SELECT EMP_ID, LAST_NAME FROM EMPLOYEE_TBL WHERE CITY = 'Seattle' ORDER BY EMP_ID;",
```

Comment:
can you add tests for queries with aliases? For example:
```
`"SELECT EMP_ID, LAST_NAME FROM EMPLOYEE_TBL AS EMPL_TBL WHERE CITY = 'Seattle' ORDER BY EMP_ID;"`
SELECT EMP_ID, LAST_NAME FROM EMPLOYEE_TBL EMPL_TBL WHERE CITY = 'Seattle' ORDER BY EMP_ID;"`
```
Below I see code `sqlparser.String(table.(*sqlparser.AliasedTableExpr)) == forbiddenTable` and not sure what actually checks - aliases or table names.

Reply:
Actually, I used this code as in the tests of sqlparser (https://github.com/xwb1989/sqlparser/blob/master/analyzer_test.go#L125)

Reply:
Hmm, thanks. It seems it doesn't work for aliased tables